lib/components/features/lista.dart

```diff
@@ -1,28 +1,135 @@
+import 'package:bytebank/models/feature.dart';
+import 'package:bytebank/webclient/features_webclient.dart';
 import 'package:flutter/material.dart';
 
 import 'item.dart';
 
-class ListaFeatures extends StatelessWidget {
+const _featureHistorico = 'historico';
+const _tituloFeatureHistorico = 'Histórico';
+
+const _featureTransferir = 'transferir';
+const _tituloFeatureTransferir = "Transferir";
+
+const _featureCartaoDeCredito = 'cartao de credito';
+const _tituloFeatureCartaoDeCredito = 'Cartão de crédito';
+
+const _featureAjuda = 'ajuda';
+const _tituloFeatureAjuda = 'Ajuda';
+
+class ListaFeatures extends StatefulWidget {
+  final Function() quandoClicaTransacoes;
+  final Function() quandoClicaContatos;
+
+  const ListaFeatures({
+    this.quandoClicaTransacoes,
+    this.quandoClicaContatos,
+  });
+
+  @override
+  State<StatefulWidget> createState() {
+    return _ListaFeaturesState();
+  }
+}
+
+class _ListaFeaturesState extends State<ListaFeatures> {
+  List<ItemFeature> _features = List();
+
+  @override
+  void initState() {
+    super.initState();
+    _buscaTodas();
+  }
+
   @override
   Widget build(BuildContext context) {
     return Padding(
       padding: const EdgeInsets.all(8.0),
       child: Container(
         height: 120,
-        child: ListView(
-          scrollDirection: Axis.horizontal,
-          children: <Widget>[
-            FeatureItem(
-              icone: Icons.monetization_on,
-              texto: 'Transferência',
-            ),
-            FeatureItem(
-              icone: Icons.people,
-              texto: 'Contatos',
-            ),
-          ],
-        ),
+        child: _tentaCarregarFeatures(),
       ),
     );
   }
+
+  Widget _tentaCarregarFeatures() {
+    if (_features.isEmpty) {
+      return Progresso();
+    }
+    return Features(_features);
+  }
+
+  _buscaTodas() async {
+    final List<Feature> features = await FeatureWebClient().todas();
+    final itemFeaturesDisponiveis = _carregaFeaturesDisponiveis(features);
+
+    //TODO código apenas para carregar padrões features caso não tenha comunicação com API
+    if (itemFeaturesDisponiveis.isNotEmpty) {
+      itemFeaturesDisponiveis.add(_devolveItemFeature(
+          Feature(_featureTransferir, true)));
+      itemFeaturesDisponiveis.add(_devolveItemFeature(
+          Feature(_featureHistorico, true)));
+    }
+
+    //TODO estou usando essa técnica para atualizar a lista, existe alguma outra abordagem recomendada?
+    setState(() {
+    _features.addAll(itemFeaturesDisponiveis);
+    });
+  }
+
+
+  List<ItemFeature> _carregaFeaturesDisponiveis(List<Feature> features) {
+    return features
+        //TODO considerei essa técnica para transformar e filtrar, tem algo mais sucinto?
+        .map((feature) => _devolveItemFeature(feature))
+        .where((itemFeature) => itemFeature != null)
+        .toList();
+  }
+
+  //TODO Tem algum pattern para usar esse tipo de solução?
```

Comment:
- Lookup tables
- Polymorphism
- Pattern matching

Matching translated text is also a bad idea...
Is there any way to introduce another more type-defining value? (simple IDs or just some enum type)

Reply:
I can adjust the Feature model putting the id, can I to use Enum as Id then? About Polymorphism and Pattern matching, do you have a sample code? I'm have problem to do it considering the callback `quandoClica()`.

Reply:
Yes, you can use the enum. Just needs a way for deserialization and it should be done.
Pattern matching might fall short because of dart's current capabilities.

I just don't get why ItemFeature has an optional `quandoClica` Could you elaborate?

Reply:
The reason for this approach is to enable the click listener for each element from `Dashboard`. Currently, the `ItemFeature` emits the click event and delegates to` Dashboard` to decide what need to do. I kept as optional, because I'll add new features by web API only to show how to create a 'feature flag' solution, so considering that new features doesn't have a screen, I did this way.

Reply:
Hmm

What I'm seeing here in this method is this:
- Verifies some data (text string in this case)
- Creates and populates a UI element: ItemFeature
- And binds the delegate calls on the callbacks

Maybe if you had the enum for the `Feature` model then you could have the UI element populate itself for example (with the texts and icons it requires), and have a callback that only returns the enum itself, then the enum->delegate logic could be somewhere else.
As to why it is optional, shouldn't the `Feature` model itself contain that information? (move that kind of responsibility outside this UI widget)